### ISSUE TITLE
netopeer2-server core dumped while calling create-subscription rpc

### ISSUE THREAD

**Issue by yangtian127** [CESNET/netopeer2#843]
Hi, Michal
When I use command 'subscribe' to subscribe notification on two netopeer-cli client, the netopeer-server core dumped, error message:
**netopeer2-server: /sysrepo-1.4.104/src/common.c:2998: sr_rwunlock: Assertion `!rwlock->readers[0] && !rwlock-upgr && (rwlock->writer == cid)' failed.
Aborted (core dumped)**

Another question: on a same netopeer-cli can execute subscribe command more than once and return OK, is this right? 

netopeer2-1.1.53
sysrepo-1.4.104

thanks

**Comment by michalvasko** [comment]
Could you please try ti with the latest *sysrepo* `devel`?

**Comment by yangtian127** [comment]
I try it with sysrepo-devel, but netopeer-server segmentation fault.
must compile with netopeer2-devel?

**Comment by yangtian127** [comment]
Rebuild netopeer2-server with netopeer2-devel and sysrepo-devel, there is no 'core dumped',
But another issue:
On a same netopeer2-cli client, create-subscribe command can be executed more than once, and same number of notifications received for one event.

**Comment by michalvasko** [comment]
Yes, you can subscribe several times and if a notification matches more of your subscriptions, you will receive it more times. Why do you think this is a problem?

**Comment by yangtian127** [comment]
RFC 5277
6.5.  Modifications to Existing Operations

   When a <create-subscription> is sent while another subscription is
   active on that session, the following error will be returned:

      Tag: operation-failed

      Error-type: protocol

      Severity: error

      Error-info: none

      Description: Request could not be completed because the requested
      operation failed for some reason not covered by any other error
      condition.

So, I think you can only subscribe once in the same session. In the old version netopeer2-server， the session was checked for subscription:

    /* check that the session is not in the current subscribers list */
    for (i = 0; i < subscribers.num; i++) {
        if (subscribers.list[i]->session == ncs) {
            /* already subscribed */
            e = nc_err(NC_ERR_IN_USE, NC_ERR_TYPE_PROT);
            nc_err_set_msg(e, "Already subscribed.", "en");
            ereply = nc_server_reply_err(e);
            goto unlock_error;
        }
    }

**Comment by michalvasko** [comment]
Right, I had remembered this was forbidden but could not find the information when I was implementing *netopeer2* so I left it allowed. Thanks, should return an error now.

**Comment by yangtian127** [comment]
There is still a problem here for BUGFIX 81eed76 :
  
  /* RFC 5277 section 6.5 */
    if (nc_session_get_notif_status(ncs)) {
        sr_set_error(session, NULL, "Session already subscribed.");
        rc = SR_ERR_EXISTS;
        goto cleanup; **// The error should be returned here, otherwise in cleanup the session's subscription status will be set to 0**
    }

 I checked the latest release and deve versions and found that they were very different, It should be a major change, I want to know when there will be a new release.

Thanks!

**Comment by michalvasko** [comment]
Yeah, you are right, should be fixed now. As for the release, I cannot tell you much, there is just too much work with a higher priority.

**Comment by adamzyg** [comment]
After using devel branch of netopeer2 and sysrepo, I also encountered this problem.

**Comment by michalvasko** [comment]
Even after the last fix or what do you mean?

**Comment by adamzyg** [comment]
> Even after the last fix or what do you mean?

```
libnetconf2: 1.1.36
libyang: 1.0.215
sysrepo: 1.4.118
netopeer2: 1.1.65
```
the first subscribe notification return `operation-failed`, but netopeer2 return `rpc ok` while sending again.

**Comment by michalvasko** [comment]
Yes, the current latest *netopeer2* version is 1.1.66, please try that one.

**Comment by adamzyg** [comment]
> Yes, the current latest _netopeer2_ version is 1.1.66, please try that one.

```
root@oru-project:~# netopeer2-cli 
get_netconf_dir: Configuration directory "/home/root/.netopeer2-cli" did not exist, created.
load_config: No saved history.
load_config: No saved configuration.
> connect 
The authenticity of the host 'localhost' cannot be established.
ssh-rsa key fingerprint is d8:87:c2:68:3e:39:24:6d:a2:4b:d5:ce:fa:27:e7:a0:b2:04:ae:0a.
Are you sure you want to continue connecting (yes/no)? yes
Interactive SSH Authentication
Type your password:
Password: 
> subscribe --filter-xpath "/o-ran-sync:ptp-state-change/ptp-state"
OK
> subscribe --filter-xpath "/o-ran-sync:ptp-state-change/ptp-state"
ERROR
	type:     protocol
	tag:      operation-failed
	severity: error
	message:  Session already subscribed.

> subscribe --filter-xpath "/o-ran-sync:ptp-state-change/ptp-state"
ERROR
	type:     protocol
	tag:      operation-failed
	severity: error
	message:  Session already subscribed.

> subscribe --filter-xpath "/o-ran-fm:alarm-notif"
ERROR
	type:     protocol
	tag:      operation-failed
	severity: error
	message:  Session already subscribed.

> subscribe --filter-xpath "/o-ran-fm:alarm-notif"
ERROR
	type:     protocol
	tag:      operation-failed
	severity: error
	message:  Session already subscribed.

> 
```
It still has issues. The first subscription success, butall subsequent new subscriptions report ·operation-failed·.

**Comment by yangtian127** [comment]
As described in RFC5177 section 6.5, you can subscribe only once in the same session, so this is the right behavior.

**Comment by adamzyg** [comment]
> As described in RFC5177 section 6.5, you can subscribe only once in the same session, so this is the right behavior.

Only the first subscription succeeds, and all other different subscriptions fail. Is this the right behavior?

**Comment by michalvasko** [comment]
That is what the RFC says, that if you already have a subscription, any other attempts at subscribing another one fail. What is not clear about that?

**Comment by adamzyg** [comment]
OK, I see.
